Native functions

nikomatsakis opened the pull request:
Initial implementation of native functions.

First commit documents the plan.

~~This PR needs a rebase, apparently, but also it doesn't work!~~

Works now!

fpoli commented on src/jvm.rs:
```diff
@@ -125,6 +133,104 @@ fn get_or_default_init_jvm() -> crate::GlobalResult<JvmPtr> {
+            let obj = result.to_java().execute_with(&mut jvm);
+            eprintln!("obj = {obj:?}");
+            match obj {
+                Ok(Some(p)) => p.as_raw().as_ptr(),
```
Here, isn't `obj: Local` deallocated on `Drop` when this Rust function terminates, while it should instead remain allocated to be used from Java? I guess that if the returned object points to a deallocated instance then the JVM might conservatively interpret it as being `null`.

nikomatsakis replied:
*Ah*, yeah, that's prob it, interesting!

nikomatsakis replied:
Good catch @fpoli, I'll tweak that

nikomatsakis commented:
Hey @fpoli -- have time to give this a review now that it's no longer in draft status?

fpoli commented on src/jvm.rs:
```diff
@@ -125,6 +133,100 @@ fn get_or_default_init_jvm() -> crate::GlobalResult<JvmPtr> {
+) -> jni_sys::jobject
+where
+    J: Upcast<crate::java::lang::Object> + Upcast<J>,
+    R: ToJavaImpl<J> + std::fmt::Debug,
```
Do you still need the `Debug`?
```suggestion
    R: ToJavaImpl<J>,
```

nikomatsakis replied:
Oh, I do not, good catch.

fpoli commented:
Is it possible to return Java objects from a native function? I tried changing the `native_greetings.rs` test to use the following native function, but it's rejected.
```rust
#[duchess::java_function(native_greeting.Native::baseGreeting)]
fn base_greeting(
    _this: &native_greeting::Native,
    name: &java::lang::String,
) -> duchess::GlobalResult<java::lang::String> {
    todo!()
}
```
The error is:
```
   --> tests/ui/native_greeting.rs:15:50
    |
15  | #[duchess::java_function(native_greeting.Native::baseGreeting)]
    |                                                  ^^^^^^^^^^^^ the trait `ToJavaImpl<duchess::java::lang::String>` is not implemented for `duchess::java::lang::String`
```

fpoli commented:
```rust
#[duchess::java_function(native_greeting.Native::baseGreeting)]
fn base_greeting(
    env: &mut JvmEnv<'_>, // Optional?
    _this: &native_greeting::Native,
    name: &java::lang::String,
) -> duchess::GlobalResult<String> {
    let _ = env;
    let name: String = name.to_rust().execute()?;
    Ok(format!("Hello, {name}"))
}
```
fails to compile with
```
error: extra argument(s) on Rust function, only 2 argument(s) are expected
  --> tests/ui/native_greeting.rs:19:5
   |
19 |     name: &java::lang::String,
   |     ^^^^
```
The documentation explains that `env` is optional, so I didn't expect this error.

fpoli commented on book/src/java_object.md:
```diff
@@ -0,0 +1 @@
+# The `JavaObject` trait
```
Why this empty page? Is it linked from elsewhere in the book?

nikomatsakis replied:
I think this trait was going to have a bigger role than it would up having...? Probably an ok page to have though.

fpoli commented on book/src/linking_native_functions.md:
````diff
@@ -0,0 +1,96 @@
+To avoid this, you can create **suites** of java functions. The idea is that the `link` method accepts both individual `JavaFunction` structs but also `Vec<JavaFunction>` suites. You can then write a function in your module that returns a `Vec<JavaFunction>` with all the java functions defined locally:
+
+```rust
+use duchess::prelude::*;
````
I think `JavaFunctions` is not in the prelude, because I tried defining a `java_functions()` in the `native_greetings.rs` test and I get the following:
```
error[E0412]: cannot find type `JavaFunction` in this scope
  --> tests/ui/native_greeting.rs:24:28
   |
24 | fn java_functions() -> Vec<JavaFunction> {
   |                            ^^^^^^^^^^^^ not found in this scope
   |
help: consider importing this struct
   |
3  | use duchess::plumbing::JavaFunction;
   |
```

nikomatsakis replied:
D'oh. Fixed. It's not in the *prelude* but it is in the crate outside of `plumbing` now.

Side note but I wonder if we can find a better way to deal with `plumbing`. It'd be cool if you could annotate something so the compiler wouldn't offer it in suggestions; or perhaps if the compiler would just avoid "doc-hidden" suggestions.

fpoli commented on book/src/linking_native_functions.md:
````diff
@@ -0,0 +1,96 @@
+You can also compose suites from other crates or modules:
+
+```rust
+fn java_functions() -> Vec<duchess::JavaFunction> {
````
For consistency with the other snippets
```suggestion
fn java_functions() -> Vec<JavaFunction> {
```

fpoli commented on book/src/linking_native_functions.md:
````diff
@@ -0,0 +1,96 @@
+```
+
+**How it works.** The call `foo::java_fn()` returns a `duchess::JavaFunction` struct. The `java_fn` method is defined in the duchess `JavaFn` trait; that trait is implemented on a struct type `foo` that is created by the `#[java_function]` decorator. This trait is in the duchess prelude, which is why you need to `use duchess::prelude::*`.
+
````
Linking the same function multiple times seems to work fine. If it's intended, should this be mentioned in the documentation?

fpoli commented on book/src/to_java.md:
```diff
@@ -0,0 +1,49 @@
+let data = format!("Hello, Duchess!");
+let hashCode: i32 =
+    data.to_java()   // Returns a `JvmOp` producing a `java::lang::String`
+        .hashCode()  // Returns a `JvmOp` invoking `hashCode` on this string
```
`public int hashCode();` is not declared for `java.lang.String` in `src/java.rs`.

Could you double check this snippet? If I try to compile it (replacing `hashCode` with `length`) I get the following error, but I'm not sure if it's because I miss some import or if it's because the chaining is still unimplemented.
```
error[E0599]: no method named `length` found for struct `ToJavaOp` in the current scope
  --> tests/ui/native_greeting.rs:21:40
   |
21 |     let nameJava: i32 = name.to_java().length().execute();
   |                                        ^^^^^^ method not found in `ToJavaOp<'_, String, String>`
```

nikomatsakis replied:
Yep. We really gotta get the snippets working to the point where they can be auto-compiled.

nikomatsakis replied:
Oh, I see the problem, there's no Deref impl for these ...

fpoli commented on java/native_greeting/Native.java:
```diff
@@ -0,0 +1,9 @@
+    }
+
+    native String baseGreeting(String name);
+}
```
```suggestion
}

```

fpoli commented on src/error.rs:
```diff
@@ -89,12 +85,5 @@ impl<'jvm> Error<Local<'jvm, Throwable>> {
 pub fn check_exception<'jvm>(jvm: &mut Jvm<'jvm>) -> Result<'jvm, ()> {
     let env = jvm.env();
     // SAFETY: we don't hold on to the return env ptr
```
```suggestion
    // SAFETY: we don't hold on to the return env ptr
```

fpoli commented on src/thread.rs:
```diff
@@ -41,6 +42,23 @@ fn attached_or(
     })
 }
 
+#[must_use = "remember to call `detach_from_jni_callback`"]
+pub unsafe fn attach_from_jni_callback(env: EnvPtr<'_>) -> State {
+    let env: EnvPtr<'static> = unsafe { std::mem::transmute(env) };
+    STATE.with(|state| state.replace(State::AttachedPermanently(env)))
+}
+
+pub unsafe fn detach_from_jni_callback(env: EnvPtr<'_>, s0: State) {
+    let env: EnvPtr<'static> = unsafe { std::mem::transmute(env) };
+    STATE.with(|state| {
+        let s1 = state.replace(s0);
+        assert!(
+            s1 == State::AttachedPermanently(env),
+            "invalid prior state `{s1:?}`"
+        );
+    });
+}
```
What's the safety condition for these two functions?

fpoli commented on src/jvm.rs:
```diff
@@ -125,6 +133,100 @@ fn get_or_default_init_jvm() -> crate::GlobalResult<JvmPtr> {
+            match obj {
+                Ok(Some(p)) => p.into_raw().as_ptr(),
+                Ok(None) => std::ptr::null_mut(),
+                Err(e) => todo!("convert from exception `{e:?}` into java exception"),
```
The `todo!` (here an inside `rust_panic_to_java_exception`) will skip the `detach_from_jni_callback`, breaking the invariant of `STATE`. I wonder how bad that is. Maybe some safety condition is violated as well.

fpoli replied:
A type that calls `detach_from_jni_callback` on `Drop` might be useful, instad of manually checking all possible unwind paths.

nikomatsakis replied:
Great suggestion! Done.

fpoli reviewed (approved):
The comment about panics in `native_function_returning_object`/`native_function_returning_scalar` is the only big one. All other comments are minor.

nikomatsakis commented:
@fpoli 

Addressed feedback:

* [x] removed Debug requirement
* [x] test for returning a java object
* [x] removed doc of `env` as it is not implemented 
* [x] inline comments

nikomatsakis commented:
Thanks @fpoli for the detailed feedback! That really helped catch a lot of oversights.
